File: List_1201/zzx/142.py
class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        c = s = f = head # 初始化快慢指针的起点和相遇点
        while f and f.next: # 快指针循环条件
            s,f = s.next,f.next.next # 快慢指针同时右移动
            if s == f: # 相遇点
                while c != s: # 寻找入环点
                    c,s = c.next,s.next # 同时移动直到相遇
                return c # 返回入环点
        return None # 未相遇，返回空

File: List_1201/jiqing/142.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

 //the difference between this and #141 is that in this q we should find the intersection node
 //the solution given below can be proved with math method.
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;       
            if(fast == slow){
                ListNode ptr = head;
                while(ptr != slow){
                    slow = slow.next;
                    ptr = ptr.next;
                }
                return ptr;
            }
        }
        return null;
    }
}